single_kernel_mongo/core/vm_workload.py

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Kubernetes workload definition."""
+
+import subprocess
+from collections.abc import Mapping
+from itertools import chain
+from logging import getLogger
+from pathlib import Path
+from shutil import copyfile
+
+from ops import Container
+from tenacity import retry, retry_if_result, stop_after_attempt, wait_fixed
+from typing_extensions import override
+
+from single_kernel_mongo.config.literals import (
+    CRON_FILE,
+    SNAP,
+    VmUser,
+)
+from single_kernel_mongo.config.models import Role
+from single_kernel_mongo.core.workload import WorkloadBase
+from single_kernel_mongo.exceptions import WorkloadExecError, WorkloadServiceError
+from single_kernel_mongo.lib.charms.operator_libs_linux.v2 import snap
+
+logger = getLogger(__name__)
+
+
+class VMWorkload(WorkloadBase):
+    """Wrapper for performing common operations specific to the MongoDB Snap."""
+
+    substrate = "vm"
+    container: None
+    users = VmUser()
+
+    def __init__(self, role: Role, container: Container | None) -> None:
+        super().__init__(role, container)
+        self.snap = SNAP
+        self.mongod = snap.SnapCache()[self.snap.name]
+
+    @property
+    @override
+    def workload_present(self) -> bool:
+        return self.mongod.present
+
+    @override
+    def start(self) -> None:
+        try:
+            self.mongod.start(services=[self.service])
+        except snap.SnapError as e:
+            logger.exception(str(e))
+            raise WorkloadServiceError(str(e)) from e
+
+    @override
+    def get_env(self) -> dict[str, str]:
+        return {self.env_var: self.mongod.get(self.snap_param)}
+
+    @override
+    def update_env(self, parameters: chain[str]):
```

Comment:
It can but:
 * snap variables are still used as environment variables
 * This interface is shared with the k8s_workload which does updates an environment
 
I think it makes sens enough to name it that way so that it abstracts what is actually done according to the substrate.

Reply:
That makes sense! can you include that in a doc string so its easy for a new contributor to understand?

Reply:
Sure, this will be in `single_kernel_mongo/core/workload.py`.